数据包括2000年到2016年美国环保署记录的加利福尼亚州四种主要污染物情况，包括二氧化氮、二氧化硫、一氧化碳和臭氧的含量， 总数超过57万。 目的在于发现空气中这四种污染物含量整体变化的趋势, 以及四种污染物对总体空气质量造成影响的差异。

In [1]:
import pandas as pd
import cufflinks as cf

cf.set_config_file(offline=True)

### 1. 加载数据

In [2]:
path = r"D:\extra\美国污染数据\California_pollution_2000_2016.csv"

In [3]:
data = pd.read_csv(path)

### 2. 查看数据集的简单描述

In [4]:
data.head()

State Code  County Code  Site Num                 Address       State  \
0           6           13         2  2956-A TREAT BOULEVARD  California   
1           6           13         2  2956-A TREAT BOULEVARD  California   
2           6           13         2  2956-A TREAT BOULEVARD  California   
3           6           13         2  2956-A TREAT BOULEVARD  California   
4           6           13         2  2956-A TREAT BOULEVARD  California   

         County     City  Date Local          NO2 Units   NO2 Mean  ...  \
0  Contra Costa  Concord  2000-01-01  Parts per billion  14.782609  ...   
1  Contra Costa  Concord  2000-01-01  Parts per billion  14.782609  ...   
2  Contra Costa  Concord  2000-01-01  Parts per billion  14.782609  ...   
3  Contra Costa  Concord  2000-01-01  Parts per billion  14.782609  ...   
4  Contra Costa  Concord  2000-01-02  Parts per billion  16.043478  ...   

           SO2 Units  SO2 Mean  SO2 1st Max Value SO2 1st Max Hour  SO2 AQI  \
0  Parts per billion  0.636364                2.0                9      3.0   
1  Parts per billion  0.636364                2.0                9      3.0   
2  Parts per billion  0.600000                2.0               11      NaN   
3  Parts per billion  0.600000                2.0               11      NaN   
4  Parts per billion  1.000000                2.0                5      3.0   

            CO Units   CO Mean  CO 1st Max Value CO 1st Max Hour  CO AQI  
0  Parts per million  0.556522               1.3               4     NaN  
1  Parts per million  0.572222               0.8               6     9.0  
2  Parts per million  0.556522               1.3               4     NaN  
3  Parts per million  0.572222               0.8               6     9.0  
4  Parts per million  0.578261               1.1              21     NaN  

[5 rows x 28 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 576142 entries, 0 to 576141
Data columns (total 28 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   State Code         576142 non-null  int64  
 1   County Code        576142 non-null  int64  
 2   Site Num           576142 non-null  int64  
 3   Address            576142 non-null  object 
 4   State              576142 non-null  object 
 5   County             576142 non-null  object 
 6   City               576142 non-null  object 
 7   Date Local         576142 non-null  object 
 8   NO2 Units          576142 non-null  object 
 9   NO2 Mean           576142 non-null  float64
 10  NO2 1st Max Value  576142 non-null  float64
 11  NO2 1st Max Hour   576142 non-null  int64  
 12  NO2 AQI            576142 non-null  int64  
 13  O3 Units           576142 non-null  object 
 14  O3 Mean            576142 non-null  float64
 15  O3 1st Max Value   576142 non-null  float64
 16  O3

In [6]:
data.describe()

State Code    County Code       Site Num       NO2 Mean  \
count    576142.0  576142.000000  576142.000000  576142.000000   
mean          6.0      53.797397    1599.569823      13.651894   
std           0.0      28.550954    2039.338943      11.081922   
min           6.0       1.000000       1.000000      -1.871429   
25%           6.0      25.000000       6.000000       5.181818   
50%           6.0      65.000000    1004.000000      10.943961   
75%           6.0      75.000000    2007.000000      19.545455   
max           6.0     111.000000    8001.000000      98.130435   

       NO2 1st Max Value  NO2 1st Max Hour        NO2 AQI        O3 Mean  \
count      576142.000000     576142.000000  576142.000000  576142.000000   
mean           25.592142         11.264417      24.110238       0.026053   
std            18.026597          7.378273      17.129925       0.011530   
min            -1.600000          0.000000       0.000000       0.000000   
25%            11.000000          6.000000      10.000000       0.017750   
50%            23.000000          9.000000      22.000000       0.026000   
75%            37.000000         19.000000      35.000000       0.034000   
max           262.000000         23.000000     131.000000       0.086214   

       O3 1st Max Value  O3 1st Max Hour         O3 AQI       SO2 Mean  \
count     576142.000000    576142.000000  576142.000000  576142.000000   
mean           0.038820        10.204474      35.722673       1.146695   
std            0.015028         3.711410      20.432386       1.406326   
min            0.000000         0.000000       0.000000      -1.521739   
25%            0.029000         9.000000      25.000000       0.157143   
50%            0.038000        10.000000      32.000000       0.760000   
75%            0.047000        11.000000      40.000000       1.556522   
max            0.140000        23.000000     218.000000      40.271429   

       SO2 1st Max Value  SO2 1st Max Hour        SO2 AQI        CO Mean  \
count      576142.000000     576142.000000  288232.000000  576142.000000   
mean            2.331772          8.508692       3.598278       0.447485   
std             3.197332          6.322809       5.195931       0.375866   
min            -1.300000          0.000000       0.000000      -0.400000   
25%             0.600000          2.000000       1.000000       0.220833   
50%             1.300000          8.000000       3.000000       0.343478   
75%             3.000000         12.000000       4.000000       0.552174   
max           192.000000         23.000000     153.000000       7.508333   

       CO 1st Max Value  CO 1st Max Hour         CO AQI  
count     576142.000000    576142.000000  288060.000000  
mean           0.769284         7.430949       7.405669  
std            0.798595         7.233509       7.183791  
min           -0.400000         0.000000       0.000000  
25%            0.300000         0.000000       3.000000  
50%            0.500000         6.000000       6.000000  
75%            1.000000        10.000000       9.000000  
max           19.900000        23.000000     201.000000

In [7]:
data.shape

(576142, 28)

### 3. 数据处理

In [8]:
data.duplicated().sum()  # 查看是否存在重复

514

In [9]:
data.drop_duplicates(inplace=True)  # 删掉完全重复的行

In [10]:
data.isnull().any() # 查看各列是否存在空值

State Code           False
County Code          False
Site Num             False
Address              False
State                False
County               False
City                 False
Date Local           False
NO2 Units            False
NO2 Mean             False
NO2 1st Max Value    False
NO2 1st Max Hour     False
NO2 AQI              False
O3 Units             False
O3 Mean              False
O3 1st Max Value     False
O3 1st Max Hour      False
O3 AQI               False
SO2 Units            False
SO2 Mean             False
SO2 1st Max Value    False
SO2 1st Max Hour     False
SO2 AQI               True
CO Units             False
CO Mean              False
CO 1st Max Value     False
CO 1st Max Hour      False
CO AQI                True
dtype: bool

In [11]:
data.dropna(inplace=True)  # 我们发现SO2 AQI和CO AQI都存在空值， 接下来删掉数据集中含空值的行， 避免对后续处理造成影响

在前面使用info()方法查看数据集各列的数据类型时， 我们发现Date Local是字符型， 为了方便后续处理， 我们将其处理成日期型

In [12]:
data['Date Local'] = pd.to_datetime(data['Date Local']) # 将Date Local字段由字符型转化为日期型

### 4. 数据分析

#### （1）四种污染物含量整体变化的趋势

In [13]:
data.head()

State Code  County Code  Site Num                 Address       State  \
1            6           13         2  2956-A TREAT BOULEVARD  California   
5            6           13         2  2956-A TREAT BOULEVARD  California   
9            6           13         2  2956-A TREAT BOULEVARD  California   
13           6           13         2  2956-A TREAT BOULEVARD  California   
17           6           13         2  2956-A TREAT BOULEVARD  California   

          County     City Date Local          NO2 Units   NO2 Mean  ...  \
1   Contra Costa  Concord 2000-01-01  Parts per billion  14.782609  ...   
5   Contra Costa  Concord 2000-01-02  Parts per billion  16.043478  ...   
9   Contra Costa  Concord 2000-01-03  Parts per billion  25.521739  ...   
13  Contra Costa  Concord 2000-01-04  Parts per billion  26.041667  ...   
17  Contra Costa  Concord 2000-01-05  Parts per billion  23.772727  ...   

            SO2 Units  SO2 Mean  SO2 1st Max Value SO2 1st Max Hour  SO2 AQI  \
1   Parts per billion  0.636364                2.0                9      3.0   
5   Parts per billion  1.000000                2.0                5      3.0   
9   Parts per billion  2.454545                4.0               13      6.0   
13  Parts per billion  1.541667                4.0               17      6.0   
17  Parts per billion  1.681818                4.0               10      6.0   

             CO Units   CO Mean  CO 1st Max Value CO 1st Max Hour  CO AQI  
1   Parts per million  0.572222               0.8               6     9.0  
5   Parts per million  0.520833               0.8              23     9.0  
9   Parts per million  1.133333               1.9              23    22.0  
13  Parts per million  1.408333               2.0               0    23.0  
17  Parts per million  0.941667               1.7              23    19.0  

[5 rows x 28 columns]

In [14]:
NO2 = data['NO2 Mean'].groupby([data['Date Local'].dt.year, data['Date Local'].dt.month]).mean() # NO2 Mean
O3 = data['O3 Mean'].groupby([data['Date Local'].dt.year, data['Date Local'].dt.month]).mean()   # O3 Mean
SO2 = data['SO2 Mean'].groupby([data['Date Local'].dt.year, data['Date Local'].dt.month]).mean() # SO2 Mean
CO = data['CO Mean'].groupby([data['Date Local'].dt.year, data['Date Local'].dt.month]).mean()   # CO Mean

In [15]:
month_mean = pd.concat([NO2, O3, SO2, CO], axis=1)

In [16]:
month_mean

NO2 Mean   O3 Mean  SO2 Mean   CO Mean
Date Local Date Local                                         
2000       1           22.685029  0.013109  1.795445  1.031469
           2           17.639380  0.021247  1.271938  0.675111
           3           16.930270  0.026984  1.562430  0.575237
           4           15.240847  0.031271  1.455959  0.536494
           5           13.672839  0.032277  1.767503  0.446947
...                          ...       ...       ...       ...
2015       12          14.140778  0.019259  0.676953  0.451681
2016       1            8.599721  0.021818  0.602615  0.384096
           2           10.088560  0.022800  0.648723  0.397666
           3            5.893201  0.029233  0.484627  0.309921
           4            6.484205  0.032489  0.802672  0.317994

[196 rows x 4 columns]

In [17]:
month_mean.iplot(subplots=True, theme='white', title='四种污染物月均浓度的变化趋势')

美国空值质量（AQI）分为6级， 具体标准定义如下：

|等级|指数范围|
|-|-|
|良好|0-50|
|中等|51-100|
|对易感人群不健康|101-150|
|不健康|151-200|
|非常不健康|201-300|
|危险|301+|

即值越大，空气质量越差。 原始数据中， 我们已经有了分别以NO2、O3、SO2和CO计算的AQI， 而最终的空气质量指数取的是这里面的最大值， 相对应的那个污染物则称为首要污染物。

In [18]:
data['AQI'] = data.apply(lambda x: max([x['NO2 AQI'], x['O3 AQI'], x['SO2 AQI'], x['CO AQI']]), axis=1)

In [19]:
def air_quality_grade(x):
    if 0<= x<= 50:
        return '良好'
    elif 51<=x<=100:
        return '中等'
    elif 101<=x<=150:
        return '对易感人群不健康'
    elif 151<=x<=200:
        return '不健康'
    elif 201<=x<=300:
        return '非常不健康'
    else:
        return '危险'

In [20]:
type_data = data.groupby([data['Date Local'].dt.year, data['Date Local'].dt.month, data['Date Local'].dt.day])['AQI'].mean().map(air_quality_grade).value_counts()

In [21]:
type_data.iplot(kind='bar', theme='white', colorscale='set1', title='2000-2016各空气质量等级天数', yTitle='天数', xTitle='空气质量等级')

#### （2）四种污染物对总体空气质量造成影响的差异

我们来查看一下四种污染物平均浓度与最终AQI的相关性

In [22]:
data.head()

State Code  County Code  Site Num                 Address       State  \
1            6           13         2  2956-A TREAT BOULEVARD  California   
5            6           13         2  2956-A TREAT BOULEVARD  California   
9            6           13         2  2956-A TREAT BOULEVARD  California   
13           6           13         2  2956-A TREAT BOULEVARD  California   
17           6           13         2  2956-A TREAT BOULEVARD  California   

          County     City Date Local          NO2 Units   NO2 Mean  ...  \
1   Contra Costa  Concord 2000-01-01  Parts per billion  14.782609  ...   
5   Contra Costa  Concord 2000-01-02  Parts per billion  16.043478  ...   
9   Contra Costa  Concord 2000-01-03  Parts per billion  25.521739  ...   
13  Contra Costa  Concord 2000-01-04  Parts per billion  26.041667  ...   
17  Contra Costa  Concord 2000-01-05  Parts per billion  23.772727  ...   

    SO2 Mean  SO2 1st Max Value  SO2 1st Max Hour SO2 AQI           CO Units  \
1   0.636364                2.0                 9     3.0  Parts per million   
5   1.000000                2.0                 5     3.0  Parts per million   
9   2.454545                4.0                13     6.0  Parts per million   
13  1.541667                4.0                17     6.0  Parts per million   
17  1.681818                4.0                10     6.0  Parts per million   

     CO Mean  CO 1st Max Value  CO 1st Max Hour CO AQI   AQI  
1   0.572222               0.8                6    9.0  25.0  
5   0.520833               0.8               23    9.0  28.0  
9   1.133333               1.9               23   22.0  38.0  
13  1.408333               2.0                0   23.0  42.0  
17  0.941667               1.7               23   19.0  36.0  

[5 rows x 29 columns]

In [23]:
columns = ['Date Local', 'NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI']

In [24]:
data[columns+['AQI']]

Date Local  NO2 AQI  O3 AQI  SO2 AQI  CO AQI   AQI
1      2000-01-01       25      25      3.0     9.0  25.0
5      2000-01-02       28      27      3.0     9.0  28.0
9      2000-01-03       38      12      6.0    22.0  38.0
13     2000-01-04       42       9      6.0    23.0  42.0
17     2000-01-05       36      19      6.0    19.0  36.0
...           ...      ...     ...      ...     ...   ...
576123 2016-04-26        8      37      1.0     2.0  37.0
576127 2016-04-27        5      39      0.0     3.0  39.0
576131 2016-04-28        6      34      0.0     2.0  34.0
576135 2016-04-29        8      35      0.0     3.0  35.0
576139 2016-04-30       12      38      0.0     3.0  38.0

[143967 rows x 6 columns]

下面计算出4中污染物的平均浓度与总体空气质量的相关系数矩阵

In [25]:
aqi_correlation_matrix = data[['NO2 Mean', 'O3 Mean', 'SO2 Mean', 'CO Mean', 'AQI']].corr()

In [26]:
aqi_correlation_matrix

NO2 Mean   O3 Mean  SO2 Mean   CO Mean       AQI
NO2 Mean  1.000000 -0.496972  0.348466  0.724479  0.376626
O3 Mean  -0.496972  1.000000 -0.120382 -0.455529  0.412828
SO2 Mean  0.348466 -0.120382  1.000000  0.346402  0.165183
CO Mean   0.724479 -0.455529  0.346402  1.000000  0.238079
AQI       0.376626  0.412828  0.165183  0.238079  1.000000

并画出对应的相关系数热力图

In [27]:
aqi_correlation_matrix.iplot(kind='heatmap', colorscale='ylgn', theme='white')

我们发现， 和总体AQI相关性最高的污染物分别是：O3(0.413)， NO2(0.377)， CO(0.238)， SO2(0.165)， 即O3和NO2为造成AQI上升的两种主要污染物

### 5. 结论

在2000-2016年间， NO2、SO2、CO三种污染物的浓度都呈下降趋势， 只有O3不降反升； 而四种污染物中， O3和NO2为造成AQI上升的两种主要污染物